### ISSUE TITLE
[D3D9] Rayman Origins: some effects are incorrectly rendered

### ISSUE THREAD

**Issue by Askannz** [doitsujin/dxvk#1875]
The "ghost links" that connect mobs to Electoon cages are incorrectly rendered with DXVK, but correctly rendered with WineD3D. I've seen similar glitches with smoke and flame effects later in the game. It's a really minor issue and does not affect the gameplay as far as I can see.

Correct (native and WineD3D):

![20210102144040_1](https://user-images.githubusercontent.com/9202863/103452497-61ce3f00-4d24-11eb-96c4-47701a9ac30d.jpg)

Incorrect (DXVK):

![20210102155056_1](https://user-images.githubusercontent.com/9202863/103452555-f33db100-4d24-11eb-9045-1126f8dac984.jpg)

### Software information
Rayman Origins (Steam version, AppID 207490)

### System information
- GPU: GTX 1060 (mobile)
- Driver: 455.45.01
- Wine version: Proton 5.13-4
- DXVK version: tested with the default Proton version and with the latest master (commit https://github.com/doitsujin/dxvk/commit/d0d1d99537011880058e4d22a6d2c56b056c4a2e)

### Apitrace file(s)
https://f002.backblazeb2.com/file/RobinLange-Filesharing/rayman.trace.7z
The issue only appears at the end of the trace, I had to go through menus and a bit of gameplay first.

### Log files

[Rayman Origins_d3d9.log](https://github.com/doitsujin/dxvk/files/5759859/Rayman.Origins_d3d9.log)

:frog:

**Comment by Askannz** [comment]
Update: just got a new AMD GPU and it looks like the issue doesn't appear with RADV:

![20210123123110_1](https://user-images.githubusercontent.com/9202863/105564953-86906200-5d78-11eb-9792-3c89b6399134.jpg)

**Comment by NerosTie** [comment]
Tested with my Nvidia 730M (460.32.03) and your trace and the issue is present. So it could be a bug with Nvidia GPUs.

**Comment by CME42** [comment]
The game tries to lerp between some number x and INF which results in NaN on nvidia and intel.
This hack works around the issue, but meh 🐸:
```
diff --git a/src/dxso/dxso_compiler.cpp b/src/dxso/dxso_compiler.cpp
index f15efd44..d2fdfdf2 100644
--- a/src/dxso/dxso_compiler.cpp
+++ b/src/dxso/dxso_compiler.cpp
@@ -2059,10 +2063,22 @@ namespace dxvk {
         }
         break;
       case DxsoOpcode::Lrp:
+        std::array<uint32_t, 3> srcTmp;
+
+        srcTmp[0] = emitRegisterLoad(src[0], mask).id;
+        srcTmp[1] = emitRegisterLoad(src[1], mask).id;
+        srcTmp[2] = emitRegisterLoad(src[2], mask).id;
+
+        if (m_moduleInfo.options.d3d9FloatEmulation) {
+          for (int i = 0; i < 3; i++)
+            srcTmp[i] = m_module.opNMin(typeId, srcTmp[i],
+            m_module.constfReplicant(FLT_MAX, result.type.ccount));
+        }
+
         result.id = m_module.opFMix(typeId,
-          emitRegisterLoad(src[2], mask).id,
-          emitRegisterLoad(src[1], mask).id,
-          emitRegisterLoad(src[0], mask).id);
+          srcTmp[2],
+          srcTmp[1],
+          srcTmp[0]);
         break;
       case DxsoOpcode::Frc:
         result.id = m_module.opFract(typeId,
```
![Bildschirmfoto von 2021-01-24 17-50-34](https://user-images.githubusercontent.com/11275913/105637373-81f9a580-5e6d-11eb-9419-ffe89c2e5b97.png)

**Comment by Joshua-Ashton** [comment]
I am more interested in how the INF got there in the first place... does it come from the constant buffer?

**Comment by CME42** [comment]
Oh right, it's coming out of a fma instruction, probably makes more sense to nmin the result of this. Seems to be working fine.
```
diff --git a/src/dxso/dxso_compiler.cpp b/src/dxso/dxso_compiler.cpp
index f15efd44..72d32842 100644
--- a/src/dxso/dxso_compiler.cpp
+++ b/src/dxso/dxso_compiler.cpp
@@ -1796,6 +1796,10 @@ namespace dxvk {
             result.id,
             emitRegisterLoad(src[2], mask).id);
         }
+        if (m_moduleInfo.options.d3d9FloatEmulation) {
+          result.id = m_module.opNMin(typeId, result.id,
+            m_module.constfReplicant(FLT_MAX, result.type.ccount));
+        }
         break;
       case DxsoOpcode::Mul:
         result.id = m_module.opFMul(typeId,
```

**Comment by Joshua-Ashton** [comment]
I mean even further back...

**Comment by dungeon007** [comment]
You cant debug this just with this trace, as it happen in other places in a game always slightly differently... anyway I was playing/testing this game years back with nine and without regression fix was:
https://cgit.freedesktop.org/mesa/mesa/commit/?id=394420ebb3eec6d8956588ee13380d5059e14bea
 Anyway try just force FLT_MAX/2 for RCP on dxvk, to fix it for else vendor drivers, like nvidia, amdvlk, whatever.... might even fix some god of rays inconsistencies in other apps on these drivers for free. 🤣

**Comment by Joshua-Ashton** [comment]
I know the cause exactly, I'd prefer to just fix it properly.

**Comment by dungeon007** [comment]
Sure thing, here comes Nine years of properness 🤣
https://bugs.winehq.org/show_bug.cgi?id=30550

**Comment by Joshua-Ashton** [comment]
Well, I more mean, writing a Vulkan extension to have proper anything * 0 behaviour.  It'll happen eventually, I just need to find time for it :)

**Comment by K0bin** [comment]
Should be fixed by c13395db970d16a5631a7c0926e2c4529d2710c3

**Comment by Sterophonick** [comment]
The Electoon hair that is supposed to extend to create a path for the player to traverse at the end of World 1 is mostly missing in DXVK. The hair renders correctly when using dgVoodoo2, however, at the cost of some performance.

## DXVK
![image](https://user-images.githubusercontent.com/34801996/158519208-843135a4-3bfa-404f-aeb8-212dbfaf2792.png)

## dgVoodoo2 (Correct)
![image](https://user-images.githubusercontent.com/34801996/158518239-806b332e-062a-42d2-8e71-fce37e08540c.png)

## System Information
GPU: Intel CometLake-U GT2 [UHD Graphics]
Driver: Mesa3D 21.3.7
Wine Version: GE-Proton7-9
DXVK version: https://github.com/doitsujin/dxvk/commit/89b1f025ebad7498ac21e5726306245ebc646e59 (async)

## Logs
APITrace: https://anonfiles.com/r9q787O7x7/Rayman_Origins.trace_7z
D3D9 Log: [Rayman Origins_d3d9.log](https://github.com/doitsujin/dxvk/files/8259021/Rayman.Origins_d3d9.log)
